### PR TITLE
Streamline driver and master infrastructure

### PR DESCRIPTION
**Pull request by ipa-cmh**
# NodeInterface
This PR will streamline the driver and master infrastructure in order to make driver and master code independent of managed or non managed node. This is done by introducing a node_canopen_driver_interface and a node_canopen_master_interface similar to what is done in rclcpp::Node and rclcpp_lifecycle. A driver or master implementation shall provide an implementation of the interface.

For example:
```
class NodeCanopenDriver : public NodeCanopenDriverInterface
```

NodeCanopenDriverInterface shall look like this:
```
namespace ros2_canopen
{
    namespace node_interfaces
    {
        class NodeCanopenDriverInterface
        {
            public:
            NodeCanopenDriverInterface(){}
            /**
             * @brief Set Master
             * 
             * Sets the Lely Canopen Master Objects needed by the driver
             * to add itself to the master.
             * 
             */
            virtual void set_master(
                std::shared_ptr<lely::ev::Executor> exec,
                std::shared_ptr<lely::canopen::AsyncMaster> master) = 0;

            /**
             * @brief Demand set Master
             * 
             * This function should ask the drivers parent container to
             * set the master objects. This should result in a call to 
             * set_master function.
             * 
             */
            virtual void demand_set_master() = 0;
            

            /**
             * @brief Initialise the driver
             * 
             * In this function the ROS interface should be created and
             * potentially necessary callback groups.
             * 
             */
            virtual void init() = 0;

            /**
             * @brief Configure the driver
             * 
             * This function should contain the configuration related things,
             * such as reading parameter data or configuration data from files.
             * 
             */
            virtual void configure() = 0;

            /**
             * @brief Activate the driver
             * 
             * This function should activate the driver, consequently it needs to start all timers and threads necessary
             * for the operation of the driver.
             * 
             */
            virtual void activate() = 0;

            /**
             * @brief Deactivate the driver
             * 
             * This function should deactivate the driver, consequently it needs to stop all timers and threads that
             * are related to the operation of the diver.
             * 
             */
            virtual void deactivate() = 0;

            /**
             * @brief Cleanup the driver
             * 
             * This function needs to clean the internal state of the driver. This means
             * all data should be deleted.
             * 
             */
            virtual void cleanup() = 0;

            /**
             * @brief Shutdown the driver
             * 
             * This function should shutdown the driver.
             * 
             */
            virtual void shutdown() = 0;

            /**
             * @brief Add the driver to master
             * 
             */
            virtual void add_to_master() = 0;

            /**
             * @brief Remove the driver from master
             * 
             */
            virtual void remove_from_master() = 0; 
        };
    }
}
```

The NodeCanopenMasterInterface shall look like this:
```
namespace ros2_canopen
{
    namespace node_interfaces
    {
        class NodeCanopenMasterInterface
        {
        public:
            /**
             * @brief Initialise Master
             *
             */
            virtual void init() = 0;

            /**
             * @brief Configure the driver
             *
             * This function should contain the configuration related things,
             * such as reading parameter data or configuration data from files.
             *
             */
            virtual void configure() = 0;

            /**
             * @brief Activate the driver
             *
             * This function should activate the driver, consequently it needs to start all timers and threads necessary
             * for the operation of the driver.
             *
             */
            virtual void activate() = 0;

            /**
             * @brief Deactivate the driver
             *
             * This function should deactivate the driver, consequently it needs to stop all timers and threads that
             * are related to the operation of the diver.
             *
             */
            virtual void deactivate() = 0;

            /**
             * @brief Cleanup the driver
             *
             * This function needs to clean the internal state of the driver. This means
             * all data should be deleted.
             *
             */
            virtual void cleanup() = 0;

            /**
             * @brief Shutdown the driver
             *
             * This function should shutdown the driver.
             *
             */
            virtual void shutdown() = 0;
            /**
             * @brief Get the master object
             * 
             * @return std::shared_ptr<lely::canopen::AsyncMaster> 
             */
            virtual std::shared_ptr<lely::canopen::AsyncMaster> get_master() = 0;
            /**
             * @brief Get the executor object
             * 
             * @return std::shared_ptr<lely::canopen::Executor> 
             */
            virtual std::shared_ptr<lely::ev::Executor> get_executor() = 0;
        };
    }
}
```
# LifecycleNode and Node implmentation
A CanopenDriverInterface and a CanopenMasterInterface shall be implemented by all LifecycleNode and Node based drivers.
The canopen_core package shall provide a generic abstract classes that implement the Interface for rclcpp::Node and rclcpp_lifecycle::LifecycleNode and for Master and Driver.
A driver implmentation can provide a rclcpp::Node or rclcpp_lifecycle::LIfecycleNode by creating a derived class and exposing it as Component.

For example:
```
class BaseDriver : public ros2_canopen::CanopenDriver
{
  std::shared_ptr<node_interfaces::NodeCanopenBaseDriver<rclcpp::Node>> node_canopen_base_driver_;

public:
  BaseDriver(rclcpp::NodeOptions node_options = rclcpp::NodeOptions())
  {
       node_canopen_base_driver_ = std::make_shared<node_interfaces::NodeCanopenBaseDriver<rclcpp::Node>>(this);
       node_canopen_driver_ = std::static_pointer_cast<node_interfaces::NodeCanopenDriverInterface>(node_canopen_base_driver_);

};
#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(ros2_canopen::BaseDriver)
```

The CanopenDriverInterface shall look like this:
```
class CanopenDriverInterface
    {
    public:
        /**
         * @brief Initialise the driver
         * 
         * This function will initialise the drivers functionalities.
         * It will be called by the device_container when the node has
         * been added to the executor.
         */
        virtual void init() = 0;

        /**
         * @brief Set the master object
         * 
         * This function will set the Canopen Master Objects that are
         * necessary for the driver to be instantiated. It will be called
         * by the device container when the init_driver service is invoked.
         * 
         * @param exec 
         * @param master 
         */
        virtual void set_master(std::shared_ptr<lely::ev::Executor> exec,
                                std::shared_ptr<lely::canopen::AsyncMaster> master) = 0;

        /**
         * @brief Get the node base interface object
         * 
         * This function shall return an rclcpp::node_interfaces::NodeBaseInterface::SharedPtr.
         * The pointer will be used to add the driver node to the executor.
         * 
         * @return rclcpp::node_interfaces::NodeBaseInterface::SharedPtr 
         */
        virtual rclcpp::node_interfaces::NodeBaseInterface::SharedPtr get_node_base_interface() = 0;

        /**
         * @brief Shutdown the driver
         * 
         * This function will shutdown the driver and will especially
         * join all threads to enable a clean shutdown.
         * 
         */
        virtual void shutdown() = 0;

        /**
         * @brief Check whether this is a LifecycleNode
         * 
         * This function provides runtime information on whether the driver
         * is a lifecycle driver or not.
         * 
         * @return true 
         * @return false 
         */
        virtual bool is_lifecycle() = 0;

        /**
         * @brief Get the node canopen driver interface object
         * 
         * This function gives access to the underlying NodeCanopenDriverInterface.
         * 
         * @return std::shared_ptr<node_interfaces::NodeCanopenDriverInterface> 
         */
        virtual std::shared_ptr<node_interfaces::NodeCanopenDriverInterface> get_node_canopen_driver_interface() = 0;

    };
```

This PR will also streamline the device_container implementation, so that only a single implementation can handle both lifecycle and non-lifecycle components. The device container will automatically check if all provided components are lifecycle components or non-lifecycle components. It will inform the user if the bus.yml specifies mixed components and throw an error.

Todos:

- [x] Create node interfaces
- [x] Create driver and master interfaces
- [x] Split of master implementation as seperate package (canopen_master_driver)
- [ ] Adapt canopen_base_driver
- [ ] Adapt canopen_proxy_driver
- [ ] Adapt canopen_402_driver
- [ ] Refactor device_container
- [ ] Adapt canopen_ros2_control

### CODE REVIEW

**Comment by ipa-cmh** [comment]
@destogl @livanov93
This PR will change some things, that will hopefully make your life easier.
Specifically, device_container now exposes registered_drivers as a map of node_id and CanopenDriverInterface. This should make casting to ProxyDriver or Cia402Driver easier.
```std::map<uint16_t, std::shared_ptr<CanopenDriverInterface>> get_registered_drivers()```

I have also create some functions that will enable type checking before static_pointer_cast, as the current implementation will not pickup on false types that cannot be cast to ProxyDriver.

You can now also ask for the node_ids of drivers with a certain type (based on string name).
```std::vector<uint16_t> get_ids_of_drivers_with_type(std::string type)```

You can now also get information about the type of driver on a certain node id (only string name).
```std::string get_driver_type(uint16_t id)```

I have adapted the current canopen_system implementation to work with the refactored drivers.
Tests seem to work.

**Comment by destogl** [comment]
Hi @ipa-cmh,

This sounds great!! I find this PR has some mixed changes, and therefore it is unnecessarily long. I see that you changed formatting (indentation) in some files. Therefore, I opened a new PR #61 with formatter we are using in ros2_control to standardize code-style.

If you like it, I propose that we merge first #61, then #62 and then format this PR or try to rebase it. Then we will also get less number of changes, and we can review the files better.

**Comment by destogl** [comment]
> This PR will streamline the driver and master infrastructure in order to make driver and master code independent of managed or non managed node. This is done by introducing a node_canopen_driver_interface and a node_canopen_master_interface similar to what is done in rclcpp::Node and rclcpp_lifecycle. A driver or master implementation shall provide an implementation of the interface.

What is the mining of the word `Node` in the class names? ROS2 Node?

**Comment by ipa-cmh** [comment]
> What is the mining of the word `Node` in the class names? ROS2 Node?

Do you mean the node_xxx.hpp in node_interfaces? Basically I took this syntax over from rclcpp. They describe common node interfaces. Basically each ROS2 driver node needs to have an object that implements node_canopen_driver_interface. Same as each rclcpp:.Node and rclcpp_lifecycle node have node_base_interface etc.

My initial plan was to pass node_base, node_services etc. to avoid templating and make node_canopen_driver_interface implementations completely independent of what type of node we are working with. The node_interfaces provided by rclcpp do not expose customary function (i.e. create_service but some more basic functions like add_service). This is not very user friendly for people who want to implement their own functionality in a node_canopen_driver_interface implementation e.g. for another CANopen Profile.  Thus I pass a pointer to the node and use templating...

Long story short, node_interfaces provide a ROS2 node with an interface e.g. node_canopen_driver_interface. So the classes themselves are not ROS2 nodes. But the name indicates that they are an implementation of a node_interface.

**Comment by ipa-cmh** [comment]
> Hi @ipa-cmh,
> 
> This sounds great!! I find this PR has some mixed changes, and therefore it is unnecessarily long. I see that you changed formatting (indentation) in some files. Therefore, I opened a new PR #61 with formatter we are using in ros2_control to standardize code-style.
> 
> If you like it, I propose that we merge first #61, then #62 and then format this PR or try to rebase it. Then we will also get less number of changes, and we can review the files better.

Sorry my bad - I just like wide indents. Yeah we could have the precommit solution.

**Comment by ipa-cmh** [comment]
@destogl 
I have removed the formatting changes in canopen_system.hpp and canopen_system.cpp as far as possible. The general changes to the api should be clearer now.

**Review by uaczr — APPROVED** [review]

**Review by livanov93 — APPROVED** [review]
@ipa-cmh Thanks for this! There are a lot of changes - still, it is readable.
@destogl let's merge #61 and #62
 I will continue work on exposing 402 profile through `ros2_control` on this PR so we don't lose time.

**Comment by livanov93** [comment]
@ipa-cmh Is [this](https://github.com/ipa-cmh/ros2_canopen/blob/8db3342bb0262c81d541ba1cfc42b3ff8dbfb28e/canopen_ros2_control/src/canopen_system.cpp#L133) still correct casting?

**Comment by ipa-cmh** [comment]
> @ipa-cmh Is [this](https://github.com/ipa-cmh/ros2_canopen/blob/8db3342bb0262c81d541ba1cfc42b3ff8dbfb28e/canopen_ros2_control/src/canopen_system.cpp#L133) still correct casting?

Yes, this still works. We could now as well introduce a type check before casting, as pointer casting "fails" silently. This could be done by device_container->get_driver_type(node_id). This will return the typeid (i.e. ros2_canopen::ProxyDriver) or something else. **I have however renamed the MotionControllerDriver to ros2_canopen::Cia402Driver for logics reasons**.

**Comment by livanov93** [comment]
@ipa-cmh What is the reason the Cia402 driver is not any more special subclass of `ProxyDriver`? Wasn't this the case before?

**Comment by livanov93** [comment]
@ipa-cmh I tested lifecycle implementation with:
- `ros2 launch canopen_tests cia402_lifecycle_setup.launch.py` 
-  `ros2 lifecycle set /cia402_device_1 configure`

The result is crashed node. Do you experience the same?

**Comment by livanov93** [comment]
> @ipa-cmh I tested lifecycle implementation with:
> 
> * `ros2 launch canopen_tests cia402_lifecycle_setup.launch.py`
> * `ros2 lifecycle set /cia402_device_1 configure`
> 
> The result is crashed node. Do you experience the same?

@ipa-cmh We need to add some documentation on lifecycle handling, although it is logical to configure and activate master before the nodes in the lifecycle implementation, some documentation on order of lifecycle transition should exist.

I was successful in running with following sequence:
- `ros2 launch canopen_tests cia402_lifecycle_setup.launch.py`
- `ros2 lifecycle set /master configure`
- `ros2 lifecycle set /master activate`
- `ros2 lifecycle set /cia402_device_1 configure`
- `ros2 lifecycle set /cia402_device_1 activate`

**Comment by ipa-cmh** [comment]
> > @ipa-cmh I tested lifecycle implementation with:
> > 
> > * `ros2 launch canopen_tests cia402_lifecycle_setup.launch.py`
> > * `ros2 lifecycle set /cia402_device_1 configure`
> > 
> > The result is crashed node. Do you experience the same?
> 
> @ipa-cmh We need to add some documentation on lifecycle handling, although it is logical to configure and activate master before the nodes in the lifecycle implementation, some documentation on order of lifecycle transition should exist.
> 
> I was successful in running with following sequence:
> 
> * `ros2 launch canopen_tests cia402_lifecycle_setup.launch.py`
> * `ros2 lifecycle set /master configure`
> * `ros2 lifecycle set /master activate`
> * `ros2 lifecycle set /cia402_device_1 configure`
> * `ros2 lifecycle set /cia402_device_1 activate`

@livanov93 
When using the service lifecycle mode, there is a node called lifecycle_manager, that you can/should use to switch lifecycle states of all nodes in correct sequence. It executes the sequence you described when you activate lifecycle_manager for all drivers listed in the bus.yml. Basically just:
```
ros2 lifecycle set /lifecycle_manager configure
ros2 lifecycle set /lifecycle_manager activate
```
I have been finding and fixing some bugs in the lifecycle_manager today, related to #60. It will in state of #60 crash on deactivate, as there was some missing error catching.

Yes, I will update the documentation - it is one of the next bigger PRs on my list.

**Comment by destogl** [comment]
> Sorry my bad - I just like wide indents. Yeah we could have the precommit solution.

No problem! We can have indents as you wish, but I would like to adjust them in the `clang-format` configuration and then apply it.

this way the PR is almost impossible to review and you lose quite a lot of traceability of changes in the git history.

**Review by destogl — APPROVED** [review]
Agree with the changes and concept!

P.S. But still think that only relevant changes should be merged and other changes mergen in a separate PR.

**Comment by ipa-cmh** [comment]
> Agree with the changes and concept!
> 
> P.S. But still think that only relevant changes should be merged and other changes mergen in a separate PR.

@destogl
Reviewed this again. there was one additional change that is really unrelated to this PR, which is renaming can_interface_name to can_interface on a repository basis. This I have as well undone. The rest of the changes are related
to the restructuring. 
The only way of splitting this up in a more understandable way would be on a per package basis. However, this would mean we
have build failure in master, which I would like to avoid.

**Comment by ipa-cmh** [comment]
@livanov93 @destogl
From your approving reviews I assume you are okay with merging this in.

I will open another PR for the renaming of can_interface_name to can_interface.

**Comment by livanov93** [comment]
> @livanov93 @destogl
> 
> From your approving reviews I assume you are okay with merging this in.
> 
> 
> 
> I will open another PR for the renaming of can_interface_name to can_interface.
> 
> 

Sure.

**Comment by livanov93** [comment]
@ipa-cmh so you suggest to use instead of [this](https://github.com/livanov93/ros2_canopen/blob/c1b73c33ca7cc81c91e35bd6dff94980aaaaf318/canopen_ros2_control/src/cia402_system.cpp#L240):
```
auto motion_controller_driver = std::static_pointer_cast<ros2_canopen::Cia402Driver>(drivers[it->first]);
```

to use something like:
```
       auto impl = std::static_pointer_cast<ros2_canopen::node_interfaces::NodeCanopen402Driver<rclcpp::Node>>(drivers[it->first]->get_node_canopen_driver_interface());

```
